org/mozilla/jss/netscape/security/x509/AlgorithmId.java

```diff
@@ -177,12 +189,88 @@ public static AlgorithmId parse(byte[] val)
     }
 
     /**
-     * Constructs a parameterless algorithm ID.
+     * Constructs an algorithm ID with a fully encoded params object
+     *
+     * @ param oid the identifier for the algorithm
+     * @ param params the fully encoded AlgorithmIdentifier Object
+     */
+
+    public AlgorithmId(ObjectIdentifier oid, AlgorithmParameters params)
+        throws IOException {
+       algid = oid;
+       algParams = params;
+
+       if(algParams != null ) {
+           this.params = new DerValue(algParams.getEncoded());
+       }
+    }
+
+    /**
+     * Constructor that takes the oid and name, so the name can be cachedf or laster use.
+     * @throws NoSuchAlgorithmException
+     * @throws IOException
+     *
+     */
+
+    public AlgorithmId(ObjectIdentifier oid, String algName) throws  IOException {
+        algid = oid;
+        cachedAlgName = algName;
+        /* Create the params if our algorithm is RSA PSS related */
+        if (algName != null && algName.contains("PSS")) {
+            try {
+                this.algParams = getPSSParams(algName);
+                this.params = new DerValue(this.algParams.getEncoded());
+            } catch (Exception e) {
+                throw new IOException(e);
+            }
+        }
+    }
+
+    /**
+     * Constructs an  algorithm ID with possible RSAPSS params.
      *
      * @param oid the identifier for the algorithm
+     * @throws NoSuchAlgorithmException
+     * @throws IOException
      */
     public AlgorithmId(ObjectIdentifier oid) {
```

Comment:
Same as above, seems like this should throw the new exceptions if at all possible.

Reply:
Done, I think :)

Reply:
OH , now I remember, I wanted to keep this the same signature, since it is heavily used, could be breaking code all over the place.